### PR TITLE
[SW-2460] Add H2O CoxPH to Algo API

### PR DESCRIPTION
**Pull request by satai**
Very Work in Progress, so much WIP

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # artifacts group
 group=ai.h2o
 # Major version of H2O release
-h2oMajorVersion=3.32.0
+h2oMajorVersion=3.33.0
```
Comment (mn-mikke) [inline gradle.properties:4]: note: Don't forget to revert this line before merge

Comment (satai) [reply gradle.properties:4]: I made this Draft not to be mergable by accident.

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OSupervisedAlgorithmWithFold.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    require(
+      getOffsetCol() == null || getOffsetCol() != getFoldCol(),
+      "Specified offset column cannot be the same as the fold column!")
+    require(
+      getWeightCol() == null || getWeightCol() != getFoldCol(),
+      "Specified weight column cannot be the same as the fold column!")
+    transformedSchema
+  }
+
+  override def fit(dataset: Dataset[_]): H2OSupervisedMOJOModel = {
```
Comment (mn-mikke) [inline ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OSupervisedAlgorithmWithFold.scala:45]: Is this method override necessary if the parent class `H2OSupervisedAlgorithm`?

File: ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OCoxPHTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    val algo = new H2OCoxPH()
+      .setStartCol("start")
+      .setStopCol("stop")
+      .setLabelCol("event")
+      .setIgnoredCols(Array("id"))
+
+    val model = algo.fit(dataset)
+    val predictions = model.transform(dataset).select("prediction")
+
+    predictions.show()
```
Comment (mn-mikke) [inline ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OCoxPHTestSuite.scala:105]: Do we need this command here?

File: py/tests/unit/with_runtime_sparkling/conftest.py [hunk 4]
```diff
@@ -111,3 +111,13 @@ def craiglistDatasetPath():
...
+
+
+@pytest.fixture(scope="module")
+def heartDatasetPath():
+    return "file://" + os.path.abspath("../examples/smalldata/coxph_test/heart.csv")
+
+
+@pytest.fixture(scope="module")
+def heartDataset(spark, heartDatasetPath):
+    return spark.read.csv(heartDatasetPath, header=True, inferSchema=True)
```
Comment (mn-mikke) [inline py/tests/unit/with_runtime_sparkling/conftest.py:123]: nit: missing newline

**Review by mn-mikke — COMMENTED** [review]
May I ask you to:
- add python tests similar to https://github.com/h2oai/sparkling-water/blob/master/py/tests/unit/with_runtime_sparkling/test_isolation_forest.py
- add tests in grid search (https://github.com/h2oai/sparkling-water/blob/master/py/tests/unit/with_runtime_sparkling/test_gridsearch.py#L63)
- update documentation and add an example with the usage of the algorithm (https://github.com/h2oai/sparkling-water/blob/master/doc/src/site/sphinx/ml/sw_isolation_forest.rst, there is a notion somewhere what algorithms are exposed and supported by gridsearch)

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OSupervisedAlgorithmWithFold.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import ai.h2o.sparkling.ml.models.H2OSupervisedMOJOModel
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.sql.Dataset
+import org.apache.spark.sql.types.StructType
+import hex.Model
+
+import scala.reflect.ClassTag
+
+abstract class H2OSupervisedAlgorithmWithFold[P <: Model.Parameters: ClassTag] extends H2OSupervisedAlgorithm[P] {
```
[the remark this answers is not in the document]
Comment (satai) [reply ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OSupervisedAlgorithmWithFold.scala]: Done

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OSupervisedAlgorithmWithFold.scala (hunk 2 above)
Comment (satai) [reply ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OSupervisedAlgorithmWithFold.scala]: You are righ, removed

File: py/tests/unit/with_runtime_sparkling/conftest.py (hunk 4 above)
Comment (satai) [reply py/tests/unit/with_runtime_sparkling/conftest.py]: added

File: ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OCoxPHTestSuite.scala (hunk 3 above)
Comment (satai) [reply ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OCoxPHTestSuite.scala]: removed

**Comment by mn-mikke** [comment]
@satai Can you also take a look at the last point:

> update documentation and add an example with the usage of the algorithm (https://github.com/h2oai/sparkling-water/blob/master/doc/src/site/sphinx/ml/sw_isolation_forest.rst, there is a notion somewhere what algorithms are exposed and supported by gridsearch)

and also update https://github.com/h2oai/sparkling-water/blob/master/doc/src/site/sphinx/ml/sw_grid.rst

File: doc/src/site/sphinx/ml/sw_coxph.rst [hunk 6]
```diff
@@ -0,0 +1,105 @@
+.. _isolation_forest:
```
Comment (mn-mikke) [inline doc/src/site/sphinx/ml/sw_coxph.rst:1]: This page is about cox ph, not about isolation forest.

File: doc/src/site/sphinx/ml/sw_coxph.rst [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+            import ai.h2o.sparkling._
+            import java.net.URI
+            val hc = H2OContext.getOrCreate()
+
+        Parse the data using H2O and convert them to Spark Frame
+
+        .. code:: scala
+
+	        import org.apache.spark.SparkFiles
```
Comment (mn-mikke) [inline doc/src/site/sphinx/ml/sw_coxph.rst:33]: nit: indentation

**Review by mn-mikke — APPROVED** [review]
small comments, otherwise LGTM. 

Can you try to make your PR green (rebase to master if need)

**Pull request merged**